include/range/v3/view/iota.hpp

```diff
@@ -374,39 +457,64 @@ namespace ranges
             auto distance_to(cursor const & that) const -> CPP_ret(difference_type)( //
                 requires detail::advanceable_<From>)
             {
-                return detail::iota_distance_(from_, that.from_);
+                const auto d = detail::iota_distance_(from_, that.from_);
+                return d / step_.value();
             }
             // Extension: see https://github.com/ericniebler/stl2/issues/613
             CPP_member
             auto distance_to(sentinel const & that) const -> CPP_ret(difference_type)( //
                 requires sized_sentinel_for<To, From>)
             {
-                return that.to_ - from_;
+                const auto d = that.to_ - from_;
+                return d / step_.value();
             }
 
         public:
             cursor() = default;
-            constexpr explicit cursor(From from)
+            constexpr explicit cursor(From from, step step)
               : from_(std::move(from))
+              , step_(std::move(step))
             {}
         };
         cursor begin_cursor() const
         {
-            return cursor{from_};
+            return cursor{from_, step_};
         }
         CPP_member
-        auto CPP_fun(end_cursor)()(const requires(same_as<To, unreachable_sentinel_t>))
+        auto CPP_fun(end_cursor)()(const requires same_as<To, unreachable_sentinel_t>)
         {
             return unreachable;
         }
         CPP_member
-        auto CPP_fun(end_cursor)()(const requires(!same_as<To, unreachable_sentinel_t>))
+        auto CPP_fun(end_cursor)()(const requires(!same_as<To, unreachable_sentinel_t> &&
+                                                  !same_as<From, To>))
+        {
+            return sentinel{to_};
+        }
+        CPP_member
+        auto CPP_fun(end_cursor)()(const requires (same_as<From, To> && !detail::iota_stridable_<From, To>))
         {
-            return detail::if_then_t<same_as<From, To>, cursor, sentinel>{to_};
+            return cursor{to_, step_};
         }
+        CPP_member
+            auto CPP_fun(end_cursor)()(const requires detail::iota_stridable_<From, To>)
+        {
+            const auto d = detail::iota_distance_(from_, to_);
+            const auto m = d % step_.value();
+            if(m == 0) {
+                return cursor{to_, step_};
+            }
+            return cursor{To(to_ - m + step_), step_};
```

Comment:
A `sentinel`-returning `end_cursor` overload that `requires detail::iota_stridable_<From, To>` is missing with this logic included. There also aren't test cases for the case when the iota view doesn't model `common_range`. Seems like `::check_equal(views::iota(1, short{9}, 3) | views::reverse, {7, 4, 1});` would fail.

Reply:
I have not thought of having a step but no upper bound - which is the reason I suppose you would need a sentinel overload.  I will think about it, might be an interesting use case, although not one I thought about or ever needed.

About your second point, it will not compile, and I am not sure whether it should

Reply:
This is not about `unreachable`, see https://godbolt.org/z/AwjtpF. The bounds aren't of the same type, so the view doesn't model `common_range`. If it doesn't compile in this branch, it must be due to the missing overload.